### DISCUSSION TITLE
Self-hosted replicated Clickhouse

### DISCUSSION THREAD

**Discussion by Tristan971** [plausible/analytics#3814] [Self-Hosted Support] [answered]
Hello,

I'm migrating our old Plausible setup to a newer one, where we use Clickhouse replication. It's mostly simple-ish, but I have a few questions for that.

---

First: Can I disable automatic DB migrations on start by Plausible?
 
I plan to do them by hand before updates, but I'm a bit concerned about forgetting one day, and the auto-migrations wreaking havoc (for example converting from ReplicatedMergeTree to non-replicated MergeTree). I use the Distributed database engine, so DDL queries get replicated with the benefits (automatic sync) and downsides (wreaking a replica wreaks all of them).

---

Second: What is the purpose of the "domains_lookup" table exactly?

Seems like the goal is some sort of domain lookup cache from the name and its structure, but since it's using the Memory engine (presumably on purpose), I especially want to know what happens if has different content in 2 subsequent queries by Plausible?

Will just cause a slightly slower event handling when querying different replicas with different table content, or will it actually cause data inconsistency somehow?

---

Third: Is it still unsafe to run multiple instances of Plausible for the same domain?

I remember trying that a while ago and running into issues (can't remember exactly, it was 2.5 years ago by now), so I was curious if that was definitely-safe nowadays (obviously, I mean assuming the same PSQL and Clickhouse databases)

---

(This is all assuming latest versions of Plausible, Clickhouse, and PGSQL)

Thanks!

**Comment by ruslandoga** [comment, marked as answer]
👋 @Tristan971 

> Can I disable automatic DB migrations on start by Plausible?

Yes.

> What is the purpose of the "domains_lookup" table exactly?

It's used in the v1 -> v2 data migration to lookup site ids for domain names.

> Is it still unsafe to run multiple instances of Plausible for the same domain?

I personally haven't tried it yet so can't say for sure.

**Reply by Tristan971** [reply]
> > Can I disable automatic DB migrations on start by Plausible?
>
> Yes.

Ehm... any hint as to how I can go about and do that? :slightly_smiling_face: 

Unless the easiest is to prepopulate the migrations table with the relevant rows so that they get skipped.

> > Is it still unsafe to run multiple instances of Plausible for the same domain?
>
> I personally haven't tried it yet so can't say for sure.

Well, guess I'll report back when I give it a try. A single instance of Plausible is light enough CPU-wise that I don't have a ridiculously big need for it, but it's nice to have HA going on.

**Reply by ruslandoga** [reply]
> Ehm... any hint as to how I can go about and do that? 🙂

You just remove it from [`command`](https://github.com/plausible/hosting/blob/1d4f9e93c5d89efbaf68c1e5903f19792d703318/docker-compose.yml#L31)

**Reply by Tristan971** [reply]
Ah, right yes I'm silly... Thanks.

**Comment by arnaudbriche** [comment]
Hi @Tristan971 

Can I ask if you managed to create use Clickhouse replication ?
If so, what was your workflow with the migrations ? Translating from elixir to SQL manually ? 
Or is there a better way ?

**Reply by Tristan971** [reply]
Yes, and it works as-expected numbers-wise.

I then:
- Created the database with the Replicated engine, which ensures that DDL statements are applied on all nodes
- Ensured the replicated table engine arguments have defaults (ie that you don’t need to supply these in table DDL queries)

This means that I can now manually run DDL updates just by replacing the table engines Xyz with their ReplicatedXyz version (there were a few tables using non-mergetree engines, but they’re largely unused so not relevant and I picked replicatedmergetree for these)

I think there’s probably some way to have clickhouse itself translate queries from non-replicated to replicated table engines in DDL queries, but I think I’ll just do things manually instead since plausible is infrequently released anyway.

**Comment by doctorpangloss** [comment]
Plausible isn't really usable without replicated ClickHouse support. Is this something you guys are considering to support? I am plagued by issues because the self-hosted code assumes that ClickHouse isn't replicated. Surely this is a bad way to use a database?

I previously sponsored but stopped when you said earlier it wasn't something you wanted to do.

**Reply by Tristan971** [reply]
> I am plagued by issues because the self-hosted code assumes that ClickHouse isn't replicated

I've had no issues with that personally, and I've been running it on replicated ClickHouse for ~6 months. The only annoying bit is that migrations need to be done by hand to ensure the right ReplicatedXyz table types as far as I know?

What issues did you encounter?

**Reply by ruslandoga** [reply]
👋 @doctorpangloss 

> The only annoying bit is that migrations need to be done by hand to ensure the right ReplicatedXyz table types

Just wanted to add that the raw migration SQL can be found here: https://github.com/plausible/analytics/blob/v2.1.4/priv/ingest_repo/structure.sql - and it's now updated before each release.

**Reply by doctorpangloss** [reply]
> The only annoying bit is that migrations need to be done by hand

I can't think of any project where I have ever needed to do that.

This is something I've tried. It's just too buggy.

What I am hearing: Plausible cloud hosted is not replicated. There can be no other explanation. So cloud users will lose data.

**Reply by Tristan971** [reply]
> I can't think of any project where I have ever needed to do that.

Agree, but this is in my opinion a fault of ClickHouse itself, fankly.

You can (and should) *force* the use of replicated everything, and disable the use of non-replicated engines. ie with profile settings:

```xml
<database_replicated_allow_only_replicated_engine>true</database_replicated_allow_only_replicated_engine>
<database_replicated_allow_replicated_engine_arguments>0</database_replicated_allow_replicated_engine_arguments>
```

But, weirdly, you can't have DDL queries automatically accounting for it, as far as I know. That's the problem. And the alternative is *every single client* ever having some kind of migrations templating at runtime to replace `MergeTree` with `ReplicatedMergeTree` or something... That's what's weird tbh.

> What I am hearing: Plausible cloud hosted is not replicated. There can be no other explanation. So cloud users will lose data.

Well, CH cloud does do auto-translation exactly for this reason (cause they're aware of how untenable it is), so Plausible the org could be using that.

But even if not, data durability doesn't come from live replicas but rather from the underlying storage being reliable, and from things like streaming replication (which do not require special tables etc at least in principle).

And I say all of this while I find this very cumbersome myself. But it's still really not Plausible's fault imo.

**Reply by ruslandoga** [reply]
> I can't think of any project where I have ever needed to do that.

The reason it works this way is that once the "base" tables are "bootstrapped" in replicated mode, the rest of migrations would continue using the replicated mode: https://github.com/plausible/analytics/blob/342c3e55138ff74e554e27b68b7e9ea3d5a149cd/lib/plausible/migration_utils.ex#L8-L10

And https://github.com/search?q=repo%3Aplausible%2Fanalytics%20%20on_cluster_statement&type=code

So once you "bootstrap" the tables once, the future migrations would use the their "replicated" versions.

---

AFAIK plausible.io is using a replicated setup :)

**Reply by Tristan971** [reply]
ON CLUSTER merely causes a query to be executed on all nodes, which is only half of the CH replication story, and ironically the one that can be fixed server-side automatically.

The Replicated *database* engine specifically causes any DDL query to implicitly be ON CLUSTERed. (But ON CLUSTER is still necessary for non-database-scoped queries like user creation/management).

Separately, *table* engines are what effectively makes replication exist at all, and that’s the part where you have no server-side way (that I know) to make things automatically « upgrade », because to enable table replication you *need* the ReplicatedXYZ table engine to be used.

that is: `CREATE TABLE xyz ON CLUSTER '{cluster}' Engine MergeTree …` would create a non-replicated table. Just N times (once per node), and which would diverge from one another over time.

at least that is my understanding so far…

**Reply by ruslandoga** [reply]
Sorry, I probably linked the wrong function. The return value from `IngestRepo.clustered_table?` is also used in conditionals like this one

https://github.com/plausible/analytics/blob/342c3e55138ff74e554e27b68b7e9ea3d5a149cd/priv/data_migrations/NumericIDs/sql/create-events-v2.sql.eex#L30-L34

**Reply by Tristan971** [reply]
Ah then all good indeed, and all queries ought to just work by default afaict… Not sure why I believed otherwise in the past 🤔

Though it remains sad that Plausible has to contort itself to make this happen instead of it being baked in at the ClickHouse level 😅

**Reply by doctorpangloss** [reply]
looking at the posthog source, they default to replicated. it should be the only option.

**Reply by ruslandoga** [reply]
This would complicate self-hosted setups significantly. Right now it can be run on a single machine, with that change at least three machines would be needed.

**Reply by Tristan971** [reply]
Technically you can use replicated table engines on a single-node ClickHouse deployment without issues. Tis what we do for CI runs for our software that uses it, and it avoids the burden of managing templating in SQL migration files

**Reply by ruslandoga** [reply]
Oh, I didn't know that. Thank you for the information!

**Reply by Tristan971** [reply]
If it helps here’s a minimal config we use, since it requires also having a 1-node zookeeper cluster* (or clickhouse keeper in this case) 

https://github.com/mangadex-pub/clickhouse-ci

*: as the replicated engines parameters are not optional
